craft_artifacts/artifacts/artifacts.py

```diff
@@ -0,0 +1,132 @@
+# This file is part of craft-artifacts.
+#
+# Copyright 2024 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Artifacts support."""
+
+import abc
+import shutil
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import cast, final
+
+from pydantic import BaseModel, PrivateAttr
+
+from craft_artifacts.constraints import UniqueList
+
+
+@dataclass
+class BaseArtifactInputDirs:
+    """Base artifact input directories.
+
+    Artifacts classes should expand this to declare their own required input dirs.
+    """
+
+    default_prime_dir: Path
+
+
+@dataclass
+class BaseArtifact(metaclass=abc.ABCMeta):
+    """Base artifact definition."""
+
+    input_dirs: BaseArtifactInputDirs
+    name: str
+    outputs: list[Path] = field(init=False, default_factory=list)
+    work_dir: Path = Path.cwd()
+    keep: bool = False
+
+    def __hash__(self) -> int:
+        return hash(self.name)
```

Comment:
Do we want different artefacts with the same name but different inputs and outputs to have the same hash? That seems counterintuitive to me.

Reply:
I see that the other way around: The name should be immutable but other properties might change, so the name is the only thing that can be used to generate the hash. 

So if a user put artifacts in a dictionary, outputs (for example) or of the artifact can be updated without fear of "losing" the artifact in the dict. 

I am not 100% this is the right approach though. Am I misunderstanding something?

Reply:
I think the way Paul is approaching this makes sense, but once we get to the implementation of the data structure holding the artifacts we should add a check to fail if multiple artifacts with the same hash are added, rather than just an overwrite like a dict normally would.

Reply:
I understand this behavior is already in place since I am now using a set to hold the artifacts in `Artifacts`. So if the user tries to add an other artifact with the same name as one already in the set, it will not be added. My concern here is that the artifact will be silently rejected and maybe we want to clearly fail to help the user diagnose why one artifact is missing.

Reply:
Thanks for the clarification @upils . It's counterintuitive, but with the explanation I do agree that it's the correct behaviour.